onnxruntime/core/framework/inline_containers.h

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <cmath>
+#include <memory>
+#include <iostream>
+#include <memory_resource>
+#include <core/common/safeint.h>
+
+#include "core/framework/allocator.h"
+
+#pragma warning(push)
+#pragma warning(disable : 4127)
+#include <absl/container/inlined_vector.h>
+#include <absl/container/flat_hash_set.h>
+#include <absl/container/flat_hash_map.h>
+#pragma warning(pop)
+
+namespace onnxruntime {
+
+// Use InlinedVector for small arrays that can fit on a stack.
+// Use TensorShapeVector for shapes.
+template <typename T, size_t N>
+using InlinedVector = absl::InlinedVector<T, N>;
+
+// InlinedHashSet and InlinedHashMap are preferred
+// hash based containers. They store their values in the
+// buckets array that is allocated in one shot. It eliminated
+// per-node new/delete calls. Proper memory estimates combined with
+// OrtDeclareAllignedStackOrAllocatedBuffer may reduce the number of needed
+// allocated to 1 or completely place it on a stack.
+template <typename T>
+using InlinedHashSet = absl::flat_hash_set<T>;
+
+template <typename K, typename V>
+using InlinedHashMap = absl::flat_hash_map<K, V>;
+
+namespace pmr {
+template <typename T, size_t N>
+using InlinedVector = absl::InlinedVector<T, N, std::pmr::polymorphic_allocator<T>>;
+
+template <typename T, typename Hash = absl::container_internal::hash_default_hash<T>, typename Eq = absl::container_internal::hash_default_eq<T>>
+using InlinedHashSet = absl::flat_hash_set<T, Hash, Eq, std::pmr::polymorphic_allocator<T>>;
+
+template <typename K, typename V,
+          typename Hash = absl::container_internal::hash_default_hash<K>,
+          typename Eq = absl::container_internal::hash_default_eq<K>>
+using InlinedHashMap = absl::flat_hash_map<K, V, Hash, Eq, std::pmr::polymorphic_allocator<std::pair<const K, V>>>;
+}  // namespace pmr
+
+// Default stack size on Windows - 320K(32-bit) or 1MB (64-bit), Linux - 8 MB
+// MacOS - main - 1MB and other threads 512Kb
+#ifdef _MSC_VER
```

Comment:
The following lines are the sort of thing that can be in the Env class which abstracts the platform specific functions.

Reply:
Agree